src/System.Private.CoreLib/shared/System/Activator.RuntimeType.cs

```diff
@@ -134,20 +134,99 @@ public static partial class Activator
+                    }
+                }
+
+                return instance;
+            }
+            else
+            {
+                if (entry.ConstructClass == null)
+                    throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, entry.RuntimeType));
+
+                T instance = (T)RuntimeTypeHandle.Allocate(entry.RuntimeType);
 
-            // This is a workaround to maintain compatibility with V2. Without this we would throw a NotSupportedException for void[].
-            // Array, Ref, and Pointer types don't have default constructors.
-            if (rt.HasElementType)
-                throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, rt));
+                try
+                {
+                    entry.ConstructClass(instance);
+                }
+                catch (Exception exception)
+                {
+                    throw new TargetInvocationException(exception);
+                }
+
+                return instance;
+            }
+        }
 
-            // Skip the CreateInstanceCheckThis call to avoid perf cost and to maintain compatibility with V2 (throwing the same exceptions).
-            return (T)rt.CreateInstanceDefaultCtor(publicOnly: true, skipCheckThis: true, fillCache: true, wrapExceptions: true);
+        private delegate void ClassConstructor<T>(T instance);
+        private delegate void StructConstructor<T>(ref T instance);
+
+        private static T CreateDelegate<T>(MethodBase methodInfo)
+            where T : Delegate
+        {
+            return (T)typeof(T)
+                .GetConstructor(new[] { typeof(object), typeof(IntPtr) })!
+                .Invoke(new object?[] { null, methodInfo.MethodHandle.GetFunctionPointer() });
+        }
+
+        private sealed class ActivatorCache<T>
+        {
+            internal static readonly ActivatorCache<T> Entry = new ActivatorCache<T>();
+
+            internal readonly RuntimeType RuntimeType;
+
+            internal readonly bool IsValueType;
+            internal readonly bool IsNotActivatable;
+
+            internal readonly ClassConstructor<T>? ConstructClass;
+            internal readonly StructConstructor<T>? ConstructStruct;
+
+            private ActivatorCache()
+            {
+                RuntimeType = (RuntimeType)typeof(T);
+                IsValueType = RuntimeType.IsValueType;
+
+                // The HasElementType check is a workaround to maintain compatibility with V2.
+                // Without this we would throw a NotSupportedException for void[].
+                // Array, Ref, and Pointer types don't have default constructors.
+                IsNotActivatable = RuntimeType.IsAbstract || RuntimeType.HasElementType;
+
+                if (IsNotActivatable)
+                    return;
+
+                ConstructorInfo runtimeCtorInfo = RuntimeType.GetConstructor(BindingFlags.Instance | BindingFlags.Public, null, Type.EmptyTypes, null)!;
```

Comment:
This will silently succeed when the constructor is private. It was throwing exception before.

Reply:
We should make sure that we have test coverage for this new paths. The test coverage for the existing code was piggy backing on the test coverage for the non-generic `Activator.CreateInstance`. This is a new duplicate path so it is no longer the case.

Reply:
> This will silently succeed when the constructor is private. It was throwing exception before.

In case of a reference type it will throw exception because of this two lines:

```csharp
                if (entry.ConstructClass == null)
                    throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, entry.RuntimeType));
```

But when `T` is a value type, it won't. Will fix it and add more tests.

Reply:
> We should make sure that we have test coverage for this new paths.

Tests are in CoreFX. So how should I write them to check all paths? The compiler doesn't allow to add a parameterless constructor to structs.

Reply:
Yes, it would be nice for these tests to live in CoreFX.

The struct with default ctor needs to be written in IL. Here is an example of how to write CoreFX tests in IL: https://github.com/dotnet/corefx/tree/master/src/System.Runtime/tests/TestModule

Reply:
I can write a struct for tests in IL and put into additional project as it done for test assemblies.

Reply:
Is it worth to use this module for test structs?

Reply:
Either reuse this one, or create a new one. Either way is fine with me.